File: apollo-api/src/main/java/com/apollographql/apollo/api/internal/SimpleResponseReader.java
package com.apollographql.apollo.api.internal;

import com.apollographql.apollo.api.CustomTypeAdapter;
import com.apollographql.apollo.api.CustomTypeValue;
import com.apollographql.apollo.api.Operation;
import com.apollographql.apollo.api.ResponseField;
import com.apollographql.apollo.api.ResponseReader;
import com.apollographql.apollo.api.ScalarType;
import com.apollographql.apollo.api.ScalarTypeAdapters;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;

public final class SimpleResponseReader implements ResponseReader {
  private final Map<String, Object> recordSet;
  private final ScalarTypeAdapters scalarTypeAdapters;
  private final Map<String, Object> variableValues;

  public SimpleResponseReader(Map<String, Object> recordSet, Operation.Variables variables, ScalarTypeAdapters scalarTypeAdapters) {
    this(recordSet, variables.valueMap(), scalarTypeAdapters);
  }

  private SimpleResponseReader(Map<String, Object> recordSet, Map<String, Object> variableValues, ScalarTypeAdapters scalarTypeAdapters) {
    this.recordSet = recordSet;
    this.variableValues = variableValues;
    this.scalarTypeAdapters = scalarTypeAdapters;
  }

  @Override public String readString(ResponseField field) {
    if (shouldSkip(field)) {
      return null;
    }

    final String value = valueFor(recordSet, field);
    return checkValue(field, value);
  }

  @Override public Integer readInt(ResponseField field) {
    if (shouldSkip(field)) {
      return null;
    }

    final BigDecimal value = valueFor(recordSet, field);
    checkValue(field, value);
    return value != null ? value.intValue() : null;
  }

  @Override public Long readLong(ResponseField field) {
    if (shouldSkip(field)) {
      return null;
    }

    final BigDecimal value = valueFor(recordSet, field);
    checkValue(field, value);
    return value != null ? value.longValue() : null;
  }

  @Override public Double readDouble(ResponseField field) {
    if (shouldSkip(field)) {
      return null;
    }

    final BigDecimal value = valueFor(recordSet, field);
    checkValue(field, value);
    return value != null ? value.doubleValue() : null;
  }

  @Override public Boolean readBoolean(ResponseField field) {
    if (shouldSkip(field)) {
      return null;
    }

    final Boolean value = valueFor(recordSet, field);
    return checkValue(field, value);
  }

  @Override public <T> T readObject(ResponseField field, ResponseReader.ObjectReader<T> objectReader) {
    if (shouldSkip(field)) {
      return null;
    }

    final Map<String, Object> value = valueFor(recordSet, field);
    checkValue(field, value);
    return value != null ? objectReader.read(new SimpleResponseReader(value, variableValues, scalarTypeAdapters)) : null;
  }

  @Override public <T> List<T> readList(ResponseField field, ListReader<T> listReader) {
    if (shouldSkip(field)) {
      return null;
    }

    final List values = valueFor(recordSet, field);
    checkValue(field, values);
    final List<T> result;
    if (values == null) {
      result = null;
    } else {
      result = new ArrayList<>();
      for (Object value : values) {
        if (value == null) {
          result.add(null);
        } else {
          final T item = listReader.read(new ListItemReader(field, value));
          result.add(item);
        }
      }
    }
    return result != null ? Collections.unmodifiableList(result) : null;
  }

  @SuppressWarnings("TypeParameterUnusedInFormals")
  @Override public <T> T readCustomType(ResponseField.CustomTypeField field) {
    if (shouldSkip(field)) {
      return null;
    }

    final Object value = valueFor(recordSet, field);
    checkValue(field, value);
    final T result;
    if (value == null) {
      result = null;
    } else {
      final CustomTypeAdapter<T> typeAdapter = scalarTypeAdapters.adapterFor(field.scalarType());
      result = typeAdapter.decode(CustomTypeValue.fromRawValue(value));
    }
    return checkValue(field, result);
  }

  @Override public <T> T readConditional(ResponseField field, ConditionalTypeReader<T> conditionalTypeReader) {
    if (shouldSkip(field)) {
      return null;
    }

    final String value = valueFor(recordSet, field);
    checkValue(field, value);
    if (value == null) {
      return null;
    } else {
      if (field.type() == ResponseField.Type.INLINE_FRAGMENT) {
        for (ResponseField.Condition condition : field.conditions()) {
          if (condition instanceof ResponseField.TypeNameCondition) {
            if (((ResponseField.TypeNameCondition) condition).typeName().equals(value)) {
              return conditionalTypeReader.read(value, this);
            }
          }
        }
        return null;
      } else {
        return conditionalTypeReader.read(value, this);
      }
    }
  }

  private boolean shouldSkip(ResponseField field) {
    for (ResponseField.Condition condition : field.conditions()) {
      if (condition instanceof ResponseField.BooleanCondition) {
        final ResponseField.BooleanCondition booleanCondition = (ResponseField.BooleanCondition) condition;
        final Boolean conditionValue = (Boolean) variableValues.get(booleanCondition.variableName());
        if (booleanCondition.inverted()) {
          // means it's a skip directive
          if (Boolean.TRUE.equals(conditionValue)) {
            return true;
          }
        } else {
          // means it's an include directive
          if (Boolean.FALSE.equals(conditionValue)) {
            return true;
          }
        }
      }
    }
    return false;
  }

  private <V> V checkValue(ResponseField field, V value) {
    if (!field.optional() && value == null) {
      throw new NullPointerException("corrupted response reader, expected non null value for " + field.fieldName());
    }

    return value;
  }

  private class ListItemReader implements ResponseReader.ListItemReader {
    private final ResponseField field;
    private final Object value;

    ListItemReader(ResponseField field, Object value) {
      this.field = field;
      this.value = value;
    }

    @Override public String readString() {
      return (String) value;
    }

    @Override public Integer readInt() {
      return ((BigDecimal) value).intValue();
    }

    @Override public Long readLong() {
      return ((BigDecimal) value).longValue();
    }

    @Override public Double readDouble() {
      return ((BigDecimal) value).doubleValue();
    }

    @Override public Boolean readBoolean() {
      return (Boolean) value;
    }

    @SuppressWarnings("TypeParameterUnusedInFormals")
    @Override public <T> T readCustomType(ScalarType scalarType) {
      final CustomTypeAdapter<T> typeAdapter = scalarTypeAdapters.adapterFor(scalarType);
      return typeAdapter.decode(CustomTypeValue.fromRawValue(value));
    }

    @SuppressWarnings("unchecked")
    @Override public <T> T readObject(ObjectReader<T> objectReader) {
      final Map<String, Object> value = (Map<String, Object>) this.value;
      return objectReader.read(new SimpleResponseReader(value, variableValues, scalarTypeAdapters));
    }

    @Override public <T> List<T> readList(ListReader<T> listReader) {
      final List values = (List) value;
      if (values == null) {
        return null;
      }

      final List<T> result = new ArrayList<>();
      for (Object value : values) {
        if (value == null) {
          result.add(null);
        } else {
          final T item = (T) listReader.read(new ListItemReader(field, value));
          result.add(item);
        }
      }
      return Collections.unmodifiableList(result);
    }
  }

  @SuppressWarnings("unchecked")
  private <T> T valueFor(Map<String, Object> map, ResponseField field) {
    return (T) map.get(field.responseName());
  }
}

File: apollo-api/src/main/java/com/apollographql/apollo/api/internal/SimpleResponseReader.kt
package com.apollographql.apollo.api.internal

import com.apollographql.apollo.api.*
import java.math.BigDecimal

class SimpleResponseReader private constructor(
    private val recordSet: Map<String, Any?>,
    private val variableValues: Map<String, Any?>,
    private val scalarTypeAdapters: ScalarTypeAdapters
) : ResponseReader {

  constructor(
      recordSet: Map<String, Any>,
      variables: Operation.Variables,
      scalarTypeAdapters: ScalarTypeAdapters
  ) : this(recordSet, variables.valueMap(), scalarTypeAdapters)

  override fun readString(field: ResponseField): String? {
    if (shouldSkip(field)) {
      return null
    }

    val value = valueFor<String>(recordSet, field)
    return checkValue(field, value)
  }

  override fun readInt(field: ResponseField): Int? {
    if (shouldSkip(field)) {
      return null
    }

    val value = valueFor<BigDecimal>(recordSet, field)
    checkValue(field, value)
    return value?.toInt()
  }

  override fun readLong(field: ResponseField): Long? {
    if (shouldSkip(field)) {
      return null
    }

    val value = valueFor<BigDecimal>(recordSet, field)
    checkValue(field, value)
    return value?.toLong()
  }

  override fun readDouble(field: ResponseField): Double? {
    if (shouldSkip(field)) {
      return null
    }

    val value = valueFor<BigDecimal>(recordSet, field)
    checkValue(field, value)
    return value?.toDouble()
  }

  override fun readBoolean(field: ResponseField): Boolean? {
    if (shouldSkip(field)) {
      return null
    }

    val value = valueFor<Boolean>(recordSet, field)
    return checkValue(field, value)
  }

  override fun <T : Any> readObject(field: ResponseField, objectReader: ResponseReader.ObjectReader<T>): T? {
    if (shouldSkip(field)) {
      return null
    }

    val value = valueFor<Map<String, Any>>(recordSet, field)
    checkValue(field, value)

    return value?.let {
      objectReader.read(SimpleResponseReader(it, variableValues, scalarTypeAdapters))
    }
  }

  @Suppress("UNCHECKED_CAST")
  override fun <T : Any> readList(field: ResponseField, listReader: ResponseReader.ListReader<T>): List<T?>? {
    if (shouldSkip(field)) {
      return null
    }

    val values = valueFor<List<*>>(recordSet, field)
    checkValue(field, values)

    return values?.map { value ->
      value?.let { listReader.read(ListItemReader(field, it)) }
    }
  }

  override fun <T : Any> readCustomType(field: ResponseField.CustomTypeField): T? {
    if (shouldSkip(field)) {
      return null
    }

    val value = valueFor<Any>(recordSet, field)
    checkValue(field, value)

    return value?.let {
      val typeAdapter = scalarTypeAdapters.adapterFor<T>(field.scalarType)
      typeAdapter.decode(CustomTypeValue.fromRawValue(it))
    }
  }

  override fun <T : Any> readFragment(field: ResponseField, objectReader: ResponseReader.ObjectReader<T>): T? {
    if (shouldSkip(field)) {
      return null
    }

    val value = valueFor<String>(recordSet, field)
    checkValue<String?>(field, value)

    return value?.let { typename ->
      field.conditions
          .firstOrNull { condition -> condition is ResponseField.TypeNameCondition && condition.typeNames.contains(typename) }
          ?.let { objectReader.read(this) }
    }
  }

  private fun shouldSkip(field: ResponseField): Boolean {
    for (condition in field.conditions) {
      if (condition is ResponseField.BooleanCondition) {
        val conditionValue = variableValues[condition.variableName] as Boolean
        if (condition.inverted) {
          // means it's a skip directive
          if (java.lang.Boolean.TRUE == conditionValue) {
            return true
          }
        } else {
          // means it's an include directive
          if (java.lang.Boolean.FALSE == conditionValue) {
            return true
          }
        }
      }
    }
    return false
  }

  private fun <V> checkValue(field: ResponseField, value: V?): V? {
    if (!field.optional && value == null) {
      throw NullPointerException("corrupted response reader, expected non null value for " + field.fieldName)
    }

    return value
  }

  private inner class ListItemReader internal constructor(
      private val field: ResponseField,
      private val value: Any
  ) : ResponseReader.ListItemReader {

    override fun readString(): String {
      return value as String
    }

    override fun readInt(): Int {
      return (value as BigDecimal).toInt()
    }

    override fun readLong(): Long {
      return (value as BigDecimal).toLong()
    }

    override fun readDouble(): Double {
      return (value as BigDecimal).toDouble()
    }

    override fun readBoolean(): Boolean {
      return value as Boolean
    }

    override fun <T : Any> readCustomType(scalarType: ScalarType): T {
      val typeAdapter = scalarTypeAdapters.adapterFor<T>(scalarType)
      return typeAdapter.decode(CustomTypeValue.fromRawValue(value))
    }

    @Suppress("UNCHECKED_CAST")
    override fun <T : Any> readObject(objectReader: ResponseReader.ObjectReader<T>): T {
      val value = this.value as Map<String, Any>
      return objectReader.read(SimpleResponseReader(value, variableValues, scalarTypeAdapters))
    }

    @Suppress("UNCHECKED_CAST")
    override fun <T : Any> readList(listReader: ResponseReader.ListReader<T>): List<T?> {
      val values = value as List<*>

      return values.map { value ->
        value?.let { listReader.read(ListItemReader(field, it)) }
      }
    }
  }

  @Suppress("UNCHECKED_CAST")
  private fun <T> valueFor(map: Map<String, Any?>, field: ResponseField): T? {
    return map[field.responseName] as T?
  }
}
